### ISSUE TITLE
Level / Scene Loader

### ISSUE THREAD

**Issue by renatoc8** [renatoc8/Quad2012#2]
The loading of a level/scene needs to be pulled out of the gamestate. I propose moving the level/scene loading logic to a separate class that implements a format-independent interface, so that it may be re-used regardless of level format (BSP, JSON, etc) and location (filesystem, transmitted over a network, in-memory, etc).

It is important to distinguish the difference between a level/scene loader and something like a scene manager. In order to properly describe this specific system, the following definitions are used:

> **Level** or **World**
> Refers to a series of entities (physical and non-physical, visible and non-visible), as they stand at the beginning of a scene. For example, a Level/World contains information about the location in which all of it's entities should be spawned, the angle that they should be start-out rotated by, as well as the initial value for all of it's properties. The terms "Level" and "World" may be used interchangeably, and for the sake of simplicity, the term "level" will be used from hereon.

> **Level Loader**
A class and/or sub-system who's only responsibility is to load a level's entity information from it's data source (generally from the filesystem, or from a server over the network), and add those entities to a scene. The location of the data source, or the format of the level is an implementation detail, and should be handled within, and not exposed to the caller and/or users of the interface.

> **Scene Manager**
A class and/or sub-system in charge of managing all aspects of a "scene".

For more up-to-date definitions for these as well as other terms, see the Wiki article titled "Terminology".

**Comment by aswyk** [comment]
Good god did not expect this email

**Comment by renatoc8** [comment]
@aswyk Do you know why it's listing you as a participant? Have you posted anything in this project in the past? You could always unsubscribe from notifications, but it's probably better to take you off the list of participants if you'd rather not be. I can't find how to do that though.

**Comment by aswyk** [comment]
I'll take a look Renato, I don't mind to be honest, I tend to stalk Shawn
every once in a while to see what he is getting into .hope all is well mate.

How is the project coming along? I'm actually moving back to South Africa
with my family and will be unemployed for a while, taking a 6 month
sabatical before I start a consultant gig. If there is anything I can poke
at I'd be happy to, but don't want to to get involved if I can't devout
time. Anyways man, was a nice surprise, hope all is well and send my
regards to Shawn when you see/talk to him.

I don't know if he can see this thread, but if not, send him this pic, it's
of my daughter, sure he'll fall off his chair.

**Comment by inkadnb** [comment]
> The Scene is also in charge of simulation playback controls, such as Play, Pause, and Reset.

This is something I'm not 100% on board with, it makes sense on the surface but also seems like it should really live a level higher like the Scene Manager. It also gives too much responsibility to the Scene itself.

For example, in the case of the editor, it be beneficial if it didn't need to directly care what scene's currently loaded when it Pauses or Resets. It could simply delegate to the SceneManager to Pause or Reset the current scene.

Furthermore, what really is the process or resetting a scene? Well the easiest way of thinking about it is destroying the current scene and recreating it, who creates scenes? The SceneManager.

**Comment by renatoc8** [comment]
I see what you're saying, the only problem I have with it is that the scene becomes essentially redundant to the EntityManager. The SceneManager **could** simply call Reset() on the currently active Scene, but the actual implementation of Reset() would be done in Scene.
I think if we take that sort of responsibility away from Scene, we might as well get rid of Scene, however, I don't nessesarily agree that "resetting" a scene should involve re-creating it; that would be the act of "re-creating" a scene, not resetting it.

**Comment by inkadnb** [comment]
@renatoc8 I think you're over complicating it if you don't think about it that way. 

By destroying and recreating the scene you completely eliminate the need for even having a "Reset" implementation on the Scene itself.

Implementing "Pause" is simply not calling time-steps, so that's easy enough, we don't even need to put that in scene, we just don't Tick the scene.

Play just resumes calling Tick.

**Comment by renatoc8** [comment]
Also, as well as Play, Pause, and Reset, I was thinking of having scene work with other things, such keeping it's own Task Scheduler instance. This would broaden the scope of Scene to include other "scene"-related things such as events, queues, etc.

Under my initial idea, it's best to think of "scene" as a scene in a movie, or a play. A scene written in a play states where all of the actors and props should be located, as well as what they should do, at what time. If you want to start a scene again, you simply move all of the actors and props to their initial states, you don't re-write the entire scene.

I think if we do take away Reset, etc, from Scene, the "scene" concept should be removed entirely, and we should move the responsibilities listed above to SceneManager, and possibly rename it to something else (?).

Do you agree?

**Comment by inkadnb** [comment]
I don't agree because a Scene doesn't necessarily need to use the EntityManager. They may just spawn in their own Entities and manage them themselves, we don't really know.

Destroying the Scene and Recreating it is the simplest approach and it removes an unnecessary responsibility from the Scene. If every Scene is required to implement Reset, Pause, Play, etc it means there's more code to manage for every scene.

**Comment by renatoc8** [comment]
The idea was that there wouldn't be different scenes. The "Scene" would be the same, but the data used to populate it would be different (eg. the specific level).

I do see how you re-create approach would overall be simpler. We could work around the issue of re-loading/parsing the actual level by creating a new Asset type for levels (not sure how I feel about that) and storing that information in AssetManager. Then you can create a new scene and it would simply load the existing level information from the AssetManager without having to actually reload the level from disk (or wherever).

If we create the new scene before destroying the previous one, the assets wouldn't be reloaded from disk since they are reference-counted.

I don't however think it would be a good idea to allow them to create entities without the EntityManager since 1) you'd now have multiple potential places that you'd need to search in to find a spawned entity, 2) you could now have multiple entities with the same name, which would make referencing them from scripts a complete nightmare, 3) it breaks the factory model.

So if we take out the timeline controls (Play, Reset, etc) from Scene, do you see any other purpose for a Scene class? Perhaps we could consolidate the aspects of Scene mentioned in the original post to just be part of SceneManager, and scrap the concept of an actual "Scene" object? A "scene" could simply be an internal construct within SceneManager that would never be interacted with individually, outside of SceneManager.

**Comment by fkperkins** [comment]
I have a question about the use of Scenes, because I agree that if you only have one Scene at a time it's basically the same as calling the EntityManager, with perhaps different built in functions to make it easier. But when you are making a complex scene, it may be helpful to be able to group entities and control them together, so that the same group can appear/disappear all together or move at the same time. Kind of like layers in a photoshop program, for organization but also functionality. In that case the play, pause, etc seems like it would belong to the EntityManager, because it would orchestrate all the groups simultaneously. 

From the discussion, it seems like the scene shouldn't have the power of play and reset, because the scene is just holding onto the current positions right? A scene would still have to call the EntityManager to get the position of an entity if you move into the future/past? I could be way off though, not sure of all the EntityManager's duties.

**Comment by renatoc8** [comment]
The EntityManager manages all live entities, if you destroy one, it gets removed from EntityManager. For example, a bomb in the level explodes, it can be removed from the scene since it no longer serves any purpose. This is a different concept than my proposed SceneManager, which manages a scene, as well as potentially the timeline, etc. The EntityManager should definitely not be in charge of the timeline, or scene-related functionality though.

Your suggestion of having groups of entities was one of the proposed use cases I listed in the original post, where you could switch the active Scene, or show/hide specific scenes, etc. In my proposal, you would just have multiple Scenes, and use them as individual "groups of entities" like you mentioned, and you could then use it in the manner that you suggested.

I'm not sure of how else that kind of functionality could be accomplished, unless we introduce some sort of "Group Name" property that you could pass as a string to SceneManager or EntityManager to perform actions on certain groups of entities with the specified group name. But I don't really like that solution, it feels really forced.

**Comment by renatoc8** [comment]
So the more I think about it, the more I see Scene and SceneManager as being redundant to the existing systems, namely EntityManager.

We just need a way to pass Level information to something and have it create the entities that are defined in that level via the EntityManager. But we probably don't need a whole system just for that.

As far as "resetting" or "recreating" the scene, we can simply destroy all entities in EntityManager, then have it recreate the entities based on the Level object.

I'm struggling to think of any functionality or value that a new Scene or SceneManager class could add.

**Comment by renatoc8** [comment]
Revised original comment to address concerns expressed in comments above.
